test/antrea/networkpolicy_controller.dl

```diff
@@ -0,0 +1,364 @@
+        var addressGroupNames: Vec<string> = vec_empty();
+        var addressGroups: Vec<AddressGroup> = vec_empty();
+
+        for (peer in peers) {
+            // A networking.NetworkPolicyPeer will either have an IPBlock or a
+            // podSelector and/or namespaceSelector set.
+            match (peer.ipBlock) {
+                Some{peerIPBlock} -> {
+                    match (toAntreaIPBlock(peerIPBlock)) {
+                        Left{err} -> {
+                            log_error("Failure processing NetworkPolicy ${np.namespace}/${np.name} IPBlock ${peerIPBlock}: ${err}")
+                        },
+                        Right{ipBlock} -> {
+                            vec_push(ipBlocks, ipBlock)
+                        }
+                    }
+                },
+                None -> {
+                    var addressGroup = createAddressGroup(peer, np);
+                    vec_push(addressGroups, addressGroup);
+                    vec_push(addressGroupNames, addressGroup.name)
+                }
+            }
+        };
+        var nppeer = NetworkPolicyPeer {
+            .addressGroups  = addressGroupNames,
+            .ipBlocks       = ipBlocks
+        };
+        (nppeer, addressGroups)
+    }
+}
+
+// toAntreaProtocol converts a v1.Protocol object to an Antrea Protocol object.
+function toAntreaProtocol(npProtocol: Option<k8s.Protocol>): Protocol =
+{
+	// If Protocol is unset, it must default to TCP protocol.
+    match (npProtocol) {
+        None -> protocolTCP(),
+        Some{proto} -> proto
+    }
+}
+
+// toAntreaServices converts a networkingv1.NetworkPolicyPort object to an
+// Antrea Service object.
+function toAntreaServices(npPorts: Vec<k8s.NetworkPolicyPort>): Vec<Service> =
+{
+	var antreaServices: Vec<Service> = vec_empty();
+	for (npPort in npPorts) {
+		var antreaService = Service {
+			.protocol   = toAntreaProtocol(npPort.protocol),
+		    .port       = match (npPort.port) {
+                None    -> None: Option<signed<32>>,
+                // TODO(abhiraut): Retrieve ports for named ports.
+                Some{p} -> Some{ k8s.intOrStringIntValue(p) }
+            }
+		};
+		vec_push(antreaServices, antreaService)
+	};
+	antreaServices
+}
+
+// toAntreaIPBlock converts a networkingv1.IPBlock to an Antrea IPBlock.
+function toAntreaIPBlock(ipBlock: k8s.IPBlock): Either<string,IPBlock> =
+{
+	// Convert the allowed IPBlock to networkpolicy.IPNet.
+	match (cidrStrToIPNet(ipBlock.cidr)) {
+        Left{err} -> Left{err},
+        Right{ipNet} -> {
+            var exceptNets: Vec<IPNet> = vec_empty();
+            var err: Option<string> = None;
+            for (exc in ipBlock.except) {
+                if is_none(err) {
+                    // Convert the except IPBlock to networkpolicy.IPNet.
+                    match (cidrStrToIPNet(exc)) {
+                        Left{e} -> {
+                            // XXX: fix this atrocity when we support the return statement.
```

Comment:
Maybe TODO(leonid) instead of XXX?
So it's clear that it's a DDlog problem and not inherited from Go.